Feat-transient-base-ingestor

Myestery opened the pull request:
## Mint Ingestor: Transient

### Functionality Supported
* Ingesting from URL: Yes
* Ingesting from Contract address: Yes
* Supported Networks: Base 

### Before you submit
- [x] Ensure your generated MintTemplate works 😄
- [x] Ensure that your code is restricted to a single folder in `src/ingestors`
- [x] Ensure that all required assets are included (e.g. ABIs)
- [x] Ensure ABIs are trimmed to include only methods (1) used in the ingestor or (2) required to mint
- [x] Ensure that all exported methods are prefixed with the name of your ingestor e.g. `myMintingPlatformGetContractDetails`
- [x] Ensure that a test exists for generating a MintTemplate that will always succeed
- [x] Ensure that your code accesses **no external resources** except those passed in the `resources` object

fix #8

Myestery commented:
Hello @christinehall 
I was able to add an ingestor for the transient network.

I noticed the mint contract was different from the nft contract too

chrismaddern commented:
Thank you for this submission!! 🥳 🙏 

We've gotten a little behind on reviewing these, but will be spending some time over the next 48 hours, so look out for some feedback soon! Thank you! 🙏

chrismaddern commented on src/ingestors/transient-base/index.ts:
```diff
@@ -0,0 +1,91 @@
+
+    mintBuilder.setName(name).setDescription(description).setFeaturedImageUrl(image);
+
+    mintBuilder.setMintInstructions({
```
Do we have access to the creator? If so, can we set the creator properties on the mintBuilder

e.g. https://github.com/floornfts/mobile-minting/blob/main/src/ingestors/fxhash/index.ts#L54

Myestery replied:
yeah we do

chrismaddern commented on src/ingestors/transient-base/index.ts:
```diff
@@ -0,0 +1,91 @@
+
+    mintBuilder.setMintInstructions({
+      chainId,
+      contractAddress,
```
The contract address on this object should be the contract address used to mint.

The address of the NFT that is created should be set as `mintBuilder.setMintOutputContract()`

There's an example here https://github.com/floornfts/mobile-minting/blob/961e571afab658fa5c330651613fe4ed9aaaf570/src/ingestors/fxhash/index.ts#L53

chrismaddern commented on src/ingestors/transient-base/index.ts:
```diff
@@ -0,0 +1,91 @@
+      contractAddress,
+      mintAddress,
+      contractMethod: 'purchase',
+      contractParams: '["address", 1, 1, address]',
```
Looking at the ABI above, is this right?

https://github.com/floornfts/mobile-minting/pull/27/files#diff-7b9dec2faf20a75b53006dc8f4835e4f6d5e131c7287e038cb46cccd657bb7daR11

Are the last properties optional? And are `nftAddress` and `recipient` supposed to both be the recipient's address?

Myestery replied:
```
 // nftAddress	address	0x8F7F96eEc76c882C327a1b927D04C87031ED5481
// 1	tokenId	uint256	1
// 2	recipient	address	0xcA4700293Bd3b48E7051116635C82546F17a61e3
// 3	numberToMint	uint256	1
// 4	presaleNumberCanMint	uint256	0
// 5	proof	bytes32[]
```

The above is a sample contract interaction pulled from the explorer for a successful purchase on transient

Myestery replied:
The nftAddress should be the address of the nft
and yes recipient address should be address of recipient. ABI says this allows for buying for others

Myestery replied:
I'm a bit unsure of what `contractParams: '["address", 1, 1, address]',` should be atm

Myestery replied:
I set it to
`["${contract.contractAddress}", 1, address, 1]`,

I believe it represents the structure better now

Myestery commented:
@chrismaddern I have made the necessary corrections, should be good now

chrismaddern commented:
Hi @Myestery 

Are you able to run the tests locally?

If you:
1. Add an ALCHEMY_API_KEY var in your local .env
2. Run `yarn test`

I've also just added support for simulating transactions to the tests `basicIngestorTests`

Can you merge the floornfts/mobile-minting branch into yours? There are several failing manual tests already, but this also indicates that the constructed payload does not execute successfully onchain:

![CleanShot 2024-07-19 at 08 30 38@2x](https://github.com/user-attachments/assets/d1a9cf01-a11c-4aed-a2a7-d69703f22228)

Let me know if you have any questions, otherwise, making the tests pass is the right next step 😊

Myestery commented:
Thanks for fixing the tests,
I'll run them now

Myestery commented:
I actually have some trouble running the test
I still get ts errors after merging main branch
<img width="1046" alt="Screenshot 2024-07-19 at 13 50 07" src="https://github.com/user-attachments/assets/f99ec68b-ce90-44a4-8aa2-44e60b6efa72">

chrismaddern commented:
All local machines and CI seem to be building it okay which is odd - I'm not 100% sure what the problem is:

**Setup**
* MacOS Sonoma
* Node 20.9.0 (we could probably use a node version lock file in the repo actually)
* The rest of the deps should be pinned in yarn.lock (I assume you used `yarn install`?)

Myestery commented:
> All local machines and CI seem to be building it okay which is odd - I'm not 100% sure what the problem is:
> 
> **Setup**
> 
> * MacOS Sonoma
> * Node 20.9.0 (we could probably use a node version lock file in the repo actually)
> * The rest of the deps should be pinned in yarn.lock (I assume you used `yarn install`?)
> 
> Do you have any global TS config that could be applying?
> 
> Ah, if you're having a hard time running `yarn install`, you may want to try removing the `.npmrc` file locally -- that's designed for building and distribution, but can interact with yarn install too.
> 
> ```shell
> git update-index --assume-unchanged .npmrc
> rm .npmrc
> ```
> 
> This will remove it locally, while preventing that change from being pushed to git

I finally got it to work using node 20

Myestery commented:
@chrismaddern The tests are passing now
I noticed switching the simulate env made my tests pass but some of fxhash to fail

<img width="1046" alt="Screenshot 2024-07-19 at 17 27 38" src="https://github.com/user-attachments/assets/33ef82ba-95a9-471b-a647-129a95d13d32">

chrismaddern commented on src/ingestors/transient-base/index.ts:
```diff
@@ -0,0 +1,114 @@
+      chainId,
+      contractAddress: mintAddress,
+      contractMethod: 'purchase',
+      contractParams: `[address, ${token_id}, address, 1, 0, []]`,
```
If you merge the latest from mobile-minting main, it will now run a simulation for you in running the tests.

This reveals that these params are incorrect.

Consulting the ABI:

```
{ internalType: 'address', name: 'nftAddress', type: 'address' },
      { internalType: 'uint256', name: 'tokenId', type: 'uint256' },
      { internalType: 'address', name: 'recipient', type: 'address' },
      { internalType: 'uint256', name: 'numberToMint', type: 'uint256' },
      { internalType: 'uint256', name: 'presaleNumberCanMint', type: 'uint256' },
      { internalType: 'bytes32[]', name: 'proof', type: 'bytes32[]' },
```
The important one is...

Param 1: This is the Address of the NFT collection, not the user

Replacing this with:
```
contractParams: `["${contractAddress}", "${token_id}", address, 1, 0, []]`,
```
fixes one part of the problem

The next is that the price wei for the test is being set to `20000000000000000`, but minting on metamask reveals it's actualy 0.00209.

![CleanShot 2024-07-20 at 15 20 15@2x](https://github.com/user-attachments/assets/24921ad9-4948-4a6e-a728-2bcca7a96a2d)


The priceWei is missing the mint fee it seems. 

Hopefully this helps get it across the line!

Myestery replied:
Thanks @chrismaddern !
I figured the fee some hours ago

However, I noticed there are about 4 contract types on Base Network that Transient supports today
- ERC721TL
- ERC1155TL
- ER7160TL (The contract for Kansas smile)
- Trace

I'll be adding support for them before the next review

Myestery replied:
All tests are passing now tho
Pls let me know if I should proceed with adding support for the remaining 3 contract types

chrismaddern replied:
Amazing, yes — support for all Base supported types would be extremely helpful. Particularly the 721TL is important (and I think most of their mints?)

chrismaddern commented:
Amazing! Seeing the same thing on my end!!!

![CleanShot 2024-07-20 at 17 53 48@2x](https://github.com/user-attachments/assets/a520a2eb-765c-4cf3-af1d-cdc061dc8ed5)

Nice work... I'll wait for the final types (they're just different minter addresses, hopefully?)  and then we can get this in and unlock the bounty!! 🥳 🚀

Myestery commented:
All done @chrismaddern 
<img width="1046" alt="Screenshot 2024-07-21 at 21 03 30" src="https://github.com/user-attachments/assets/e5c7d8d9-b147-4a9f-8575-bbdf46ebcaf1">

I added more tests to ensure the other contract types are working and fixed some minor flaws with the ingestor.
I think its completed !

chrismaddern commented:
Amazing!

Could you add example URLs for each to the `successUrls` for the basicIngestorTests?

Currently only the one URL is being simulated in tests:
![CleanShot 2024-07-22 at 09 10 19@2x](https://github.com/user-attachments/assets/9ab9456e-30cc-45f5-a9f3-53a48e5a45d6)

Also, you can run simulations during testing locally by adding the following to your .env:

```
SIMULATE_DURING_TESTS=true
```

When I do this locally:
```
basicIngestorTests(new TransientIngestor(), resources, {
    successUrls: [
      'https://www.transient.xyz/stacks/kansas-smile',
      'https://www.transient.xyz/stacks/volumina-8',
      'https://www.transient.xyz/stacks/16384',
    ],
```

![CleanShot 2024-07-22 at 09 13 48@2x](https://github.com/user-attachments/assets/0bdac68a-6621-48cc-8c94-2f4e696e3c33)


If you need to fix the blockHash at which the tests run the simulation you can do so here:
https://github.com/floornfts/mobile-minting/blob/main/test/ingestors/prohibition-daily.test.ts#L27

Note that all test cases currently need to be able to simulate successfully at the same blockhash (i.e. have been live & available at the same time)

Myestery commented:
> Amazing!
> 
> Could you add example URLs for each to the `successUrls` for the basicIngestorTests?
> 
> Currently only the one URL is being simulated in tests: ![CleanShot 2024-07-22 at 09 10 19@2x](https://private-user-images.githubusercontent.com/1068437/350976314-9ab9456e-30cc-45f5-a9f3-53a48e5a45d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NTQ1ODAsIm5iZiI6MTcyMTY1NDI4MCwicGF0aCI6Ii8xMDY4NDM3LzM1MDk3NjMxNC05YWI5NDU2ZS0zMGNjLTQ1ZjUtYTlmMy01M2E0OGU1YTQ1ZDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MjJUMTMxODAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzM3YTUzNDdmNWYwOTc5ZDcxNWFjMWJkMGFkZTYyNzQyYmNkNjQ1MDViMTYyYzg3NWY0ZjdmY2JlZjdlNWM3MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.TBuDZgXu71sebb0cM6f4ZjD6-WfWO87n3lYMqiL2cTw)
> 
> Also, you can run simulations during testing locally by adding the following to your .env:
> 
> ```
> SIMULATE_DURING_TESTS=true
> ```
> 
> When I do this locally:
> 
> ```
> basicIngestorTests(new TransientIngestor(), resources, {
>     successUrls: [
>       'https://www.transient.xyz/stacks/kansas-smile',
>       'https://www.transient.xyz/stacks/volumina-8',
>       'https://www.transient.xyz/stacks/16384',
>     ],
> ```
> 
> ![CleanShot 2024-07-22 at 09 13 48@2x](https://private-user-images.githubusercontent.com/1068437/350977270-0bdac68a-6621-48cc-8c94-2f4e696e3c33.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NTQ1ODAsIm5iZiI6MTcyMTY1NDI4MCwicGF0aCI6Ii8xMDY4NDM3LzM1MDk3NzI3MC0wYmRhYzY4YS02NjIxLTQ4Y2MtOGM5NC0yZjRlNjk2ZTNjMzMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MjJUMTMxODAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGMyZGIzM2Y0MGRiYzVhZWMzMTQ2ODkyZmYxYzFmNDJkYWJjZDk4OTBlNGI5MWZiYzM1ODFhNTNlNDA4ZjVkMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.98y9t-0GVBx1Y9AaGgeaW66JC2wAEMXXvVMYB56QY10)
> 
> If you need to fix the blockHash at which the tests run the simulation you can do so here: https://github.com/floornfts/mobile-minting/blob/main/test/ingestors/prohibition-daily.test.ts#L27
> 
> Note that all test cases currently need to be able to simulate successfully at the same blockhash (i.e. have been live & available at the same time)

I fixed the tests, I noticed a slight change with the ABI for [ERC721TL and ERC7160TL contracts](https://github.com/floornfts/mobile-minting/pull/27/commits/9437dc54b8c52de16561ebd1ebf9e65361d83a78)

Myestery commented:
All done 💯 
<img width="1046" alt="Screenshot 2024-07-22 at 15 16 47" src="https://github.com/user-attachments/assets/30a93cbd-3637-47c3-86de-617f23eac582">

chrismaddern commented:
Huzzah! Will review for merge.

Meanwhile, I'll DM you on the Twitter profile on your Github page with follow up for the bounty 🙂

chrismaddern commented:
@Myestery did you get the DM? 👀

Myestery commented:
> @Myestery did you get the DM? 👀

yes, seen now

Myestery commented:
> > @Myestery did you get the DM? 👀
> 
> yes, seen now

Filled.
Thanks

Myestery commented:
<img width="1308" alt="Screenshot 2024-07-25 at 00 20 33" src="https://github.com/user-attachments/assets/0dcd2384-de29-4091-9549-b893f5a4ebaa">

Hey @chrismaddern 
I noticed the test failed recently and I did some investigation
I found out that they could close the nft sale before the public sale date ended
that was the issue for kansas-smile

Myestery commented:
> <img alt="Screenshot 2024-07-25 at 00 20 33" width="1308" src="https://private-user-images.githubusercontent.com/49923152/351906831-0dcd2384-de29-4091-9549-b893f5a4ebaa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE4NjM2NzksIm5iZiI6MTcyMTg2MzM3OSwicGF0aCI6Ii80OTkyMzE1Mi8zNTE5MDY4MzEtMGRjZDIzODQtZGUyOS00MDkxLTk1NDktYjg5M2Y1YTRlYmFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI0VDIzMjI1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM1Nzc2ODMxMjJkNjYwOTE5MDE2OWE0MTA0ZmIzMDQ1YmZhZGVmNWNhNmViNzE0MjNkNmU2NzA0OWM0ZDM5NzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.W9IOdaWYDU_WHUM319u29DMJ3AIBvQ1-mbanKnWTjJU">
> Hey @chrismaddern I noticed the test failed recently and I did some investigation I found out that they could close the nft sale before the public sale date ended that was the issue for kansas-smile

It got closed some minutes ago

chrismaddern commented:
@Myestery will observed!

I added a blockhash to the basicIngestor tests in this PR here → https://github.com/floornfts/mobile-minting/pull/31

chrismaddern commented:
You should be able to merge main now! 😊
